### PR TITLE
MEN-4271: Resume downloads upon error

### PR DESCRIPTION
**Pull request by lluiscampos**
Using HTTP Range requests.

The implementation details borrow heavily from Mender golang client,
namely update_resumer.go file.

### CODE REVIEW

File: src/mender/client/deployments.py [hunk 1]
```diff
@@ -86,39 +97,151 @@ def request(
...
+                        chunk_size=1024 * 1024
+                    ):  # 1MiB at a time
+                        if not data:
+                            break
+                        fh.write(data)
+                        offset += len(data)
+                        fh.flush()
+                # Download completed in one go, return
+                log.debug(f"Got EOF. Wrote {offset} bytes. Total is {content_length}.")
+                if offset >= content_length:
```
Comment (oleorhagen) [inline src/mender/client/deployments.py:218]: How can it be bigger?

File: src/mender/client/deployments.py [hunk 2]
```diff
@@ -86,39 +97,151 @@ def request(
...
+            requests.Timeout,
+        ) as e:
+            log.error(e)
+            log.debug(f"Got Error. Wrote {offset} bytes. Total is {content_length}.")
+
+        # Prepare for next attempt
+        next_atempt_in = get_exponential_backoff_time(
+            tried, DOWNLOAD_RESUME_MAX_INTERVAL
+        )
+        if next_atempt_in is None:
```
Comment (oleorhagen) [inline src/mender/client/deployments.py:234]: `s/atempt/attempt`
`if not next_attempt_in:`

File: src/mender/client/deployments.py [hunk 3]
```diff
@@ -86,39 +97,151 @@ def request(
...
+            requests.ConnectionError,
+            requests.URLRequired,
+            requests.TooManyRedirects,
+            requests.Timeout,
+        ) as e:
+            log.error(e)
+            log.debug(f"Got Error. Wrote {offset} bytes. Total is {content_length}.")
+
+        # Prepare for next attempt
+        next_atempt_in = get_exponential_backoff_time(
```
Comment (oleorhagen) [inline src/mender/client/deployments.py:231]: Question:
Can't just `get_exponential_backoff_time` raise the error?

File: tests/unit/test_deployments.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+class CustomContentLengthResponse(Response):
+    automatically_set_content_length = False
+
+
+class TestDeploymentsMisc:
+    def test_exponential_backoff_time(self):
+        # Test with one minute maximum interval.
+        intvl = deployments.get_exponential_backoff_time(0, 1 * 60)
+        assert intvl is not None
```
Comment (oleorhagen) [inline tests/unit/test_deployments.py:35]: I think:
`assert intvl` is enough

File: tests/unit/test_deployments.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        m = deployments.header_range_regex.match("bites 123-456/789")
+        assert m is None
+        m = deployments.header_range_regex.match("bytes 1a23-456/789")
+        assert m is None
+        m = deployments.header_range_regex.match("bites 123/456/789")
+        assert m is None
+
+
+@pytest.fixture(scope="class")
+def tweak_download_resume_intervals(request):
```
Comment (oleorhagen) [inline tests/unit/test_deployments.py:114]: Nice!

File: tests/unit/test_deployments.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+
+@pytest.mark.usefixtures("tweak_download_resume_intervals")
+class TestDeploymentsDownloadAndResume:
+
+    raw_data_len = 10 * 1024 * 1024  # 10 MiB
+    raw_data = bytearray(os.urandom(10 * 1024 * 1024))
+
+    def do_run_test(self, httpserver):
+        d_dict = dict(
```
Comment (oleorhagen) [inline tests/unit/test_deployments.py:135]: Is it not already a dict?

File: tests/unit/test_deployments.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+            h_partial_response
+        )
+        httpserver.expect_ordered_request("/foobar").respond_with_handler(
+            h_full_response
+        )
+
+        self.do_run_test(httpserver)
+
+    def test_yes_range_support(self, httpserver):
+        partial_data_len = int(self.raw_data_len * 20 / 100)
```
Comment (oleorhagen) [inline tests/unit/test_deployments.py:185]: So this does floating point automatically? :thinking:

File: tests/unit/test_deployments.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+            h_partial_response
+        )
+        httpserver.expect_ordered_request("/foobar").respond_with_handler(
+            h_resume_response
+        )
+        httpserver.expect_request("/foobar").respond_with_data(
+            bytes(), status=requests.codes.OK
+        )
+
+        try:
```
Comment (oleorhagen) [inline tests/unit/test_deployments.py:284]: You could use this: https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions

**Review by oleorhagen — APPROVED** [review]
Really really nice job!

Only nitpicks and some comment points added :)

File: src/mender/client/deployments.py [hunk 9]
```diff
@@ -86,39 +97,151 @@ def request(
     return deployment_info
 
 
+def get_exponential_backoff_time(tried: int, max_interval: int) -> Optional[int]:
+    per_internal_attempts = 3
+    smallest_unit = DOWNLOAD_RESUME_MIN_INTERVAL
+
+    interval = smallest_unit
+    next_internal = interval
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/mender/client/deployments.py]: Fun fact, I did catch this one up, then used some Spacemacs commands to rewrite it, got something unexpected, undo, and then forgot about it...

Thanks

File: src/mender/client/deployments.py [hunk 10]
```diff
@@ -86,39 +97,151 @@ def request(
...
+        )
+
+    new_offset = int(match.group(1))
+    log.debug(
+        f"Successfully parsed '{h_range_str}', new_offset {new_offset}, offset {offset},"
+    )
+
+    if new_offset > offset:
+        raise DeploymentDownloadFailed(
+            f"Got a Content-Range missing bytes: '{h_range_str}'"
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/mender/client/deployments.py]: I wrote
```suggestion
            f"Missing data. Got Content-Range header: '{h_range_str}'"
```

File: src/mender/client/deployments.py [hunk 11]
```diff
@@ -86,39 +97,151 @@ def request(
...
+        )
+
+    if new_offset < offset:
+        log.debug(f"Discarding {offset-new_offset} bytes")
+        for _ in response.iter_content(chunk_size=offset - new_offset):
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/mender/client/deployments.py]: You are right, I don't know the internals.

Actually, I wanted to do `response.content = response.content[offset - new_offset : ]` but `.content` is immutable, so I came to this solution.

I'll wrap it into a 1MiB on a time read loop.

File: src/mender/client/deployments.py [hunk 12]
```diff
@@ -86,39 +97,151 @@ def request(
...
+        pass
+
+    # Loop  will try/except until download is complete or exhaust the retries
+    offset = 0
+    content_length = None
+    tried = 0
+    while True:
+        try:
+            req_headers: Dict[str, str] = {}
+            if content_length is not None:
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/mender/client/deployments.py]: Generally speaking, I personally find it more readable with "is not None" or "is None", as it is explicit what are you comparing it against. A block like if something, that something could be `-1` and still evaluate to `True`, for example.

Anyway, I modified it here and in the other occurrences to follow your guidance.

File: src/mender/client/deployments.py [hunk 13]
```diff
@@ -86,39 +97,151 @@ def request(
...
+                req_headers["Range"] = f"bytes={offset}-"
+                log.debug(f"Request with headers {req_headers}")
+            with requests.get(
+                update_url,
+                headers=req_headers,
+                stream=True,
+                verify=server_certificate if server_certificate else True,
+            ) as response:
+                if content_length is None:
+                    content_length = int(str(response.headers.get("Content-Length")))
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/mender/client/deployments.py:208]: Answer:
It is an string, but without the explicit conversion `mypy` complained. `headers` is this special case insensitive dict and might be that `mypy` cannot figure out the type.

File: src/mender/client/deployments.py (hunk 1 above)
Comment (lluiscampos) [reply src/mender/client/deployments.py:227]: Briefly discussed on the call.

In theory, a server could send more data than claims in a "Content-Length" header. I don't know if `requests` will throw an exception for it (I know for a fact that the opposite, sending less data than advertised goes unnoticed by `requests`).

Edit: I experimented a bit and added a new test case for this.

File: src/mender/client/deployments.py (hunk 3 above)
Comment (lluiscampos) [reply src/mender/client/deployments.py]: Good idea, thanks

File: tests/unit/test_deployments.py (hunk 4 above)
Comment (lluiscampos) [reply tests/unit/test_deployments.py]: Actually, as we have changed this, it cannot be `None` anymore but raising an `Exception`. I'll fix the test too.

File: tests/unit/test_deployments.py (hunk 8 above)
Comment (lluiscampos) [reply tests/unit/test_deployments.py]: Thanks!

File: tests/unit/test_deployments.py (hunk 7 above)
Comment (lluiscampos) [reply tests/unit/test_deployments.py:174]: What do you mean? I don't understand your question/concern.

In Python 3 all number literals are float, so I want to cast it as `int` before using it as array index.

File: src/mender/client/deployments.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply src/mender/client/deployments.py]: haha, so spacemacs not saving the day!

File: src/mender/client/deployments.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply src/mender/client/deployments.py]: These are only suggestions, so do what you find the best.

But this is what has been done throughout the project, so I think it's fair

File: tests/unit/test_deployments.py (hunk 7 above)
Comment (oleorhagen) [reply tests/unit/test_deployments.py:174]: Oh, really? TIL!

File: src/mender/client/deployments.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply src/mender/client/deployments.py]: Agreed. Consistency across the project above personal taste!

File: tests/unit/test_deployments.py (hunk 7 above)
Comment (lluiscampos) [reply tests/unit/test_deployments.py:174]: This is one of the most important gotchas from Python 2 to Python 3 :smile:

**Pull request merged**